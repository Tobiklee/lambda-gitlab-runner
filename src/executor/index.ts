import * as path from 'path';
import { DockerImageCode, DockerImageFunction } from 'aws-cdk-lib/aws-lambda';
import { DockerImageFunctionProps } from 'aws-cdk-lib/aws-lambda/lib/image-function';
import { Construct } from 'constructs';

/**
 * Selection of premade docker files
 */
export enum DockerFile {
  NODE_14 = 'node14.Dockerfile'
}

export interface LambdaRunnerConfiguration {
  /**
   * Gitlab - Url  from gitlab installation.
   * @example gitlab.com
   *
   * Is only used if --context gitlab-url is undefined.
   */
  readonly gitlabUrl?: string;

  /**
   * Runner token needed connect to gitlab instance.
   *
   * Is only used if --context runner-token is undefined.
   */
  readonly runnerToken?: string;

  /**
   * Select a default docker file.
   */
  readonly dockerFile?: DockerFile;

  /**
   * Additional function props. Overwrite defaults.
   *
   * You probably want to use your own docker image instead of the default one.
   * Use the default docker image as a reference.
   */
  readonly functionProps?: DockerImageFunctionProps;
}

export class LambdaRunner extends Construct {
  constructor(scope: Construct, id: string, props?: LambdaRunnerConfiguration) {
    super(scope, id);

    const gitlabUrl = this.node.tryGetContext('gitlab-url') ?? props?.gitlabUrl;
    const runnerToken = this.node.tryGetContext('runner-token') ?? props?.runnerToken;

    if (!gitlabUrl || !runnerToken) {
      throw new Error('Context variable [gitlab-url] or [runner-token] is missing.');
    }

    new DockerImageFunction(this, 'LambdaGitlabExecutor', {
      functionName: 'LambdaGitlabExecutor',
      code: DockerImageCode.fromImageAsset(path.join(__dirname), { file: props?.dockerFile ?? DockerFile.NODE_14 }),
      memorySize: 4096,
      ...props?.functionProps,
      environment: {
        GITLAB_URL: gitlabUrl,
        RUNNER_TOKEN: runnerToken,
        ...props?.functionProps?.environment,
      },
    });

  }
}
