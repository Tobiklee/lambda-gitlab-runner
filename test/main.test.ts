import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaRunner } from '../src';

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'test');

  stack.node.setContext('gitlab-url', 'gitlab.com');
  stack.node.setContext('runner-token', 'Token123');
  new LambdaRunner(stack, 'TestRunner');

  Template.fromStack(stack);
});
