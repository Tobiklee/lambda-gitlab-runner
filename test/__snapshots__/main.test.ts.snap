// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestRunnerLambdaGitlabExecutor4A806816": Object {
      "DependsOn": Array [
        "TestRunnerLambdaGitlabExecutorServiceRole328DF485",
      ],
      "Properties": Object {
        "Code": Object {
          "ImageUri": Object {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:fde3c8e6d29d7bde4e623a93720f0aca29831ff2554bd1a2d64e5bfdcbb3e2e9",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "GITLAB_URL": "gitlab.com",
            "RUNNER_TOKEN": "Token123",
          },
        },
        "FunctionName": "LambdaGitlabExecutor",
        "MemorySize": 4096,
        "PackageType": "Image",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestRunnerLambdaGitlabExecutorServiceRole328DF485",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestRunnerLambdaGitlabExecutorServiceRole328DF485": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
